File: noNonsense/01_two_pointers/active_recall.java
// --- TWO POINTERS ACTIVE RECALL ---

public class TwoPointers {
    // Type A: Opposite Ends
    public int[] oppositeEnds(int[] arr, int target) {
        int left = ___, right = ___;
        while (___ < ___) {
            int curr = arr[___] + arr[___];
            if (curr == target) return new int[]{___, ___};
            else if (curr < target) ___++;
            else ___--;
        }
        return new int[]{};
    }

    // Type B: Reader-Writer
    public int readerWriter(int[] arr) {
        int writer = ___;
        for (int reader = 0; reader < arr.length; reader++) {
            if (shouldKeep(arr[reader])) {
                int temp = arr[writer];
                arr[writer] = arr[reader];
                arr[reader] = temp;
                ___++;
            }
        }
        return writer;
    }
}

File: noNonsense/01_two_pointers/active_recall.py
# --- TWO POINTERS ACTIVE RECALL ---

# Type A: Opposite Ends
# Use: Sorted arrays, Palindromes
def opposite_ends(arr, target):
    left, right = ___, ___
    while ___ < ___:
        curr = arr[___] + arr[___]
        if curr == target:
            return [___, ___]
        elif curr < target:
            ___ += 1
        else:
            ___ -= 1
    return []

# Type B: Reader-Writer (Same Direction)
# Use: In-place removal, Moving elements
def reader_writer(arr):
    writer = ___
    for reader in range(len(arr)):
        if condition(arr[reader]):
            arr[writer], arr[reader] = arr[reader], arr[writer]
            ___ += 1
    return writer
